Controller/Index/Callback.php

```diff
@@ -0,0 +1,322 @@
+use SwedbankPay\Payments\Helper\Config as ConfigHelper;
+use SwedbankPay\Payments\Helper\Service as ServiceHelper;
+use SwedbankPay\Payments\Model\ResourceModel\OrderRepository;
+
+/**
+ * Class Callback
+ *
+ * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
+class Callback extends PaymentActionAbstract implements CsrfAwareActionInterface
+{
+    /**
+     * @var RequestContentInterface;
+     */
+    protected $requestContent;
+
+    /**
+     * @var JsonFactory
+     */
+    protected $jsonResultFactory;
+
+    /**
+     * @var OrderHelper
+     */
+    protected $orderHelper;
+
+    /**
+     * @var OrderRepositoryInterface
+     */
+    protected $magentoOrderRepo;
+
+    /**
+     * @var OrderRepository
+     */
+    protected $orderRepository;
+
+    /**
+     * @var ServiceHelper
+     */
+    protected $serviceHelper;
+
+    /**
+     * Callback constructor.
+     * @param Context $context
+     * @param JsonFactory $resultJsonFactory
+     * @param EventManager $eventManager
+     * @param ConfigHelper $configHelper
+     * @param Logger $logger
+     * @param ServiceHelper $serviceHelper
+     * @param RequestContentInterface $requestContent
+     * @param JsonFactory $jsonResultFactory
+     * @param OrderHelper $orderHelper
+     * @param OrderRepositoryInterface $magentoOrderRepo
+     * @param OrderRepository $orderRepository
+     * @SuppressWarnings(PHPMD.ExcessiveParameterList)
+     */
+    public function __construct(
+        Context $context,
+        JsonFactory $resultJsonFactory,
+        EventManager $eventManager,
+        ConfigHelper $configHelper,
+        Logger $logger,
+        ServiceHelper $serviceHelper,
+        RequestContentInterface $requestContent,
+        JsonFactory $jsonResultFactory,
+        OrderHelper $orderHelper,
+        OrderRepositoryInterface $magentoOrderRepo,
+        OrderRepository $orderRepository
+    ) {
+        parent::__construct($context, $resultJsonFactory, $eventManager, $configHelper, $logger);
+
+        $this->jsonResultFactory = $jsonResultFactory;
+        $this->orderHelper = $orderHelper;
+        $this->requestContent = $requestContent;
+        $this->magentoOrderRepo = $magentoOrderRepo;
+        $this->orderRepository = $orderRepository;
+        $this->serviceHelper = $serviceHelper;
+
+        $this->setEventName('callback');
+        $this->setEventMethod([$this, 'updatePaymentData']);
+    }
+
+    /**
+     * @return array|bool|ResponseInterface|ResultInterface|string
+     * @throws Exception
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     */
+    public function updatePaymentData()
+    {
+        $this->logger->debug(
+            sprintf(
+                "Callback controller is called with request: \n %s",
+                $this->requestContent->getContent()
+            )
+        );
+
+        $requestData = json_decode($this->requestContent->getContent(), true);
```

Comment:
In this line, we're basically decoding the json that we're getting from the callback. And then we're processing it with the SDK. 

Since it's the module that is receiving the callback, that's why we're decoding it at this level, logging it & making sure it's valid.

Reply:
Ok, but that parsing should be done by a class in the SDK. The plugin itself should delegate all low-level HTTP operations and JSON parsing to the SDK.